include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+   * the parameter @p ids specifies the place at which to insert the @p sub_tree.
+   * This function allows to combine different timer trees in a modular way.
+   * If a non empty string @p new_name is provided, the id of @p sub_tree is
+   * replaced by @p new_name when inserted into the tree.
+   */
+  void
+  insert(std::vector<std::string> const         ids,
+         std::shared_ptr<TimerTree const> const sub_tree,
+         std::string const                      new_name = "")
+  {
+    AssertThrow(ids.size() > 0, ExcMessage("Empty ID specified."));
+    AssertThrow(id == ids[0], ExcMessage("Invalid ID specified."));
+
+    std::vector<std::string> remaining_id = erase_first(ids);
+
+    bool found = false;
+    if (remaining_id.size() > 0)
+      {
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            if ((*it)->id == remaining_id[0])
+              {
+                (*it)->insert(remaining_id, sub_tree, new_name);
+                found = true;
+              }
+          }
+      }
+
+    if (found == false)
+      {
+        AssertThrow(
+          remaining_id.size() == 0,
+          ExcMessage(
+            "Subtree can not be inserted since the specified ID does not exist in this tree."));
+
+        std::shared_ptr<TimerTree> new_tree(new TimerTree());
+        new_tree->copy_from(sub_tree);
+        if (!new_name.empty())
+          new_tree->id = new_name;
+
+        // Make sure that the new tree does not already exist
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            AssertThrow(
+              new_tree->id != (*it)->id,
+              ExcMessage(
+                "Subtree can not be inserted since the tree already contains a subtree with the same ID."));
+          }
+
+        sub_trees.push_back(new_tree);
+      }
+  }
+
+  /**
+   * Prints wall time of all items of a tree without an analysis of
+   * the relative share of the children.
+   */
+  void
+  print_plain(ConditionalOStream const &pcout) const
+  {
+    unsigned int const length = get_length();
+
+    pcout << std::endl;
+
+    do_print_plain(pcout, 0, length);
+  }
+
+  /**
+   * This is the actual function of interest of this class, i.e., an
+   * analysis of wall times with a hierarchical formatting of results.
+   * Relative wall times are printed for all children of a sub-tree if
+   * a wall time has been set for the parent of these children. In this
+   * case, an additional item `other` is created in order to give insights
+   * to which extent the code has been covered with timers and to which
+   * extend time is spent in other code paths that are currently not
+   * covered by timers. The parameter @p level specifies the level for which
+   * to print results, where a value of 0 corresponds to the high level module.
+   */
+  void
+  print_level(ConditionalOStream const &pcout, unsigned int const level) const
```

Comment:
You restrict yourself unnecessarily by hard-coding `ConditionalOStream` for the first argument. I would use `template <typename OStreamType>`

Reply:
done here  https://github.com/dealii/dealii/pull/11939/commits/5581aacf5ec7433717b86a0675a0e679cbe9d40f